Added support for Backward indexscan in ORCA

Sanath97 opened the pull request:
Fixes https://github.com/greenplum-db/gpdb/issues/14730

Currently ORCA doesn't support Backward IndexScan and always assumes the index scan direction to be forward. With this commit, added support for Backward IndexScan which will be picked whenever sort and nulls direction of columns specified in ORDER BY clause is commutative to the Index's keys sort and nulls direction. For instance, consider below case.

```
create table foo (a int, b int, c int) distributed by (a);
create index foo_b on foo using btree(b);
insert into foo select i, i, i from generate_series(1,10000) i;
analyze foo;
explain select * from foo order by b desc limit 1;
```
Before this commits, ORCA generates below plan with a Sort

```
postgres=# explain select * from foo order by b desc limit 1;
                                 QUERY PLAN                   
------------------------------------------------
 Limit
   ->  Gather Motion 3:1  (slice1; segments: 3)
         Merge Key: b
         ->  Limit
               ->  Sort
                     Sort Key: b DESC
                     ->  Seq Scan on foo
 Optimizer: Pivotal Optimizer (GPORCA)
(8 rows)
```
With this commit, it will use a Backward IndexScan and eliminates the sort.

```
                   QUERY PLAN                   
------------------------------------------------
 Limit
   ->  Gather Motion 3:1  (slice1; segments: 3)
         Merge Key: b
         ->  Limit
               -> Index Scan Backward using foo_b on foo
 Optimizer: Pivotal Optimizer (GPORCA)
(8 rows)
```

chrishajas commented:
Haven't taken a look yet, but just wondering how comprehensive this is. Does this handle all known cases, or are there certain cases we're not handling? I'm not trying to increase the scope, just making sure we're not unintentionally missing things we wanted to support with this.

I see that we're not doing index-only scans here, is that something we also want to handle? What about dynamic index/index-only scans, for partitioned tables?

Sanath97 commented:
> 

As part of this change, we are handling IndexScan and IndexOnlyScans both of them. But, we are not handling Dynamic scans (both Dynamic Index and Dynamic IndexOnly) because they don't provide support for ScanDirection yet . Checked nodeDynamicIndexscan.c  looks like we are not yet supporting ScanDirection, however nodeIndexscan.c and nodeIndexOnlyscan.c do support that. 

We are supporting IndexOnlyScan as well, added scan direction parameter to CPhysicalIndexOnlyScan operator, modified IndexGet2IndexOnlyScan xform, also added tests to ensure that plan with IndexOnlyScan with Backward scan direction is being generated.

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogical.h:
```diff
@@ -32,6 +32,14 @@ namespace gpopt
+{
+	EisdBackward = 0,
+	EisdForward,
+	EisdNoMovement,
```
Why do we need EisdNoMovement?

gpopt replied:
EBackwardScan, EForwardScan

Sanath97 replied:
I couldn't think of a case we might need EisdNoMovement, but I've added that and named with prefix `Eisd` to keep them consistent with enum `EdxlIndexScanDirection`.

gpopt replied:
It is not needed to continue using Hungarian, which we tool a lot of effort to remove.

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformUtils.h:
```diff
@@ -186,7 +186,9 @@ class CXformUtils
-		BOOL indexForOrderBy = false);
+		BOOL indexForOrderBy = false,
+		EIndexScanDirection indexScanDirection =
+			EisdForward /*Default direction if unspecified*/);
```
Can we refrain from using default param?

vraghavan78 commented:
> Haven't taken a look yet, but just wondering how comprehensive this is. Does this handle all known cases, or are there certain cases we're not handling? I'm not trying to increase the scope, just making sure we're not unintentionally missing things we wanted to support with this.
> 
> I see that we're not doing index-only scans here, is that something we also want to handle? What about dynamic index/index-only scans, for partitioned tables?

+1. Similar to @dgkimura tests that enabled us to find all the other scenarios to be fixed. Please spend some time writing different kinds of tests. Similar to https://github.com/greenplum-db/gpdb/blob/main/src/test/regress/sql/gp_covering_index.sql. It did not fix everything but enabled us to figure out the landspace.

pobbatihari commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1060,24 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+            index_key_cols_sort_order->Append(GPOS_NEW(mp) ULONG(0));
+            index_key_cols_nulls_order->Append(GPOS_NEW(mp) ULONG(0));
+            ULONG rel_indoption = index_rel->rd_indoption[i];
+            if (rel_indoption & 2) {
```
can you add comments here

pobbatihari commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1060,24 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+            ULONG rel_indoption = index_rel->rd_indoption[i];
+            if (rel_indoption & 2) {
+                // Nulls first
+                //index_key_cols_nulls_order->Append(GPOS_NEW(mp) ULONG(1));
```
Remove this line if not required

pobbatihari commented:
Can you please add description for the newly added mdp files.

dgkimura reviewed (commented):
I haven't finished reviewing, but posting a few comments for now.

chrishajas commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1058,36 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+	nulls_direction = GPOS_NEW(mp) ULongPtrArray(mp);
+
+	// No other index type except B-tree support sort and nulls direction for key columns.
+	if (index_type == IMDIndex::EmdindBtree)
```
It'd also be nice to leverage amcanbackward rather than hardcoding btree here. But not sure if that might be too much work and scope creep here

Sanath97 replied:
Made the change.

khannaekta commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +197,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
-		// 2. NULLs are not Last in the specified Order by Clause.
-		if (pos->Ent(i) != COrderSpec::EntLast ||
-			!CColRef::Equals(pos->Pcr(i), (*pdrgpcrIndexColumns)[i]))
+		// Index is not applicable if either Order By Column do not match with index key
```
Comment update=>  Index is not applicable if ~either~ Order By Column do not match with index key

khannaekta commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +197,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+		// If the derived, required sort directions and nulls directions are not equal or not commutative, then the index is not applicable.
+		if (!(req_sort_direction->Equals(derived_sort_direction) &&
+			  req_nulls_direction->Equals(derived_nulls_direction)) &&
+			!(FAreIndicesCommutative(req_sort_direction, derived_sort_direction,
+									 i) &&
+			  FAreIndicesCommutative(req_nulls_direction,
+									 derived_nulls_direction, i)))
```
do we have a test where the sort and nulls direction are equal but not commutative?

Sanath97 replied:
Yes we have a test where sort and nulls direction are commutative both in unit tests and regress tests

khannaekta commented on src/backend/gporca/data/dxl/minidump/Backward-IndexScan-OrderBy-on-Multi-IndexCol.mdp:
```diff
@@ -0,0 +1,464 @@
+            create index index_dbea on test_index_with_sort_directions_on_orderby_limit using btree(d, b desc, e nulls first, a desc);
+            ANALYZE test_index_with_sort_directions_on_orderby_limit;
+
+            Query: explain select * from test_index_with_orderby_limit order by b, d limit 10;
```
query table needs update to `test_index_with_sort_directions_on_orderby_limit`

khannaekta replied:
Can you also check the query and the plan; they don't seem to match

Sanath97 replied:
Fixed it.

khannaekta commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +197,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+		// If the derived, required sort directions and nulls directions are not equal or not commutative, then the index is not applicable.
+		if (!(req_sort_direction->Equals(derived_sort_direction) &&
+			  req_nulls_direction->Equals(derived_nulls_direction)) &&
+			!(FAreIndicesCommutative(req_sort_direction, derived_sort_direction,
```
in the function `FAreIndicesCommutative()`, we iterate through all the keys in the BitVector from 0->i for each order by column. Is that really necessary, don't we just need to check for the i'th bit instead ?

Sanath97 replied:
I think it is necessary to traverse all the keys from 0 to i. For instance, if an index has multiple key columns(assume 3) and order by also same number of columns, and first order by columns directions is equal to index keys directions and next two are commutative, there is the possibility that we might return index applicable but index is not applicable as entire index isn't either equal or commutative. 

Consider the below example where 

```
create table foo(a int, b int, c int, d int);
create index foo_bcd on foo using btree(b,c,d);
```

Query is 
```
explain select b,c,d from foo order by b ,c desc,d desc limit 3;
```

This query says index would be applicable if we just check the i'th bit. But that's not the case. Hope this answers your question. Please correct me if I'm wrong. Thanks!

Sanath97 commented on src/test/regress/expected/qp_indexscan_optimizer.out:
```diff
@@ -459,15 +457,15 @@ select a from test_index_with_orderby_limit order by a NULLS FIRST limit 10;
          ->  Limit
                ->  Sort
                      Sort Key: a DESC NULLS LAST
-                     ->  Index Scan using index_a on test_index_with_orderby_limit
```
This was an example where we wanted to test a case with order by direction desc and null last. Since the table `test_index_with_orderby_limit` doesn't have any indices that match this order by cols it is now using SeqScan. Before this PR,  we didn't consider sort directions to determine if IndexScan is applicable or not and chose IndexScan just for every case where NULLS are LAST. With this PR, that is fixed. Hope this answers your question.

Sanath97 replied:
Yes, IndexScan doesn't help for this query since, the order by columns do not match with any of the b-tree indices, and that is why we used Seq Scan

Sanath97 commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,167 @@ set optimizer_enable_indexscan to off;
+ANALYZE test_partition_table;
+
+explain (costs off) select a from test_partition_table order by a limit 3;
+select a from test_partition_table order by a limit 3;
```
Added theses. Thanks for the suggestions.

gpopt commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1058,36 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+			nulls_direction->Append(GPOS_NEW(mp) ULONG(0));
+			// indoption value represents sort and nulls direction using 2 bits
+			ULONG rel_indoption = index_rel->rd_indoption[i];
+			// First bit being set represents Sort DESC and unset represents Sort ASC
+			if (rel_indoption & 1)
```
Agree with what @dgkimura said. We'd better consider using Postgres macros or constants if possible.

Sanath97 replied:
Used Macros instead of literals

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDynamicIndexGet.h:
```diff
@@ -110,6 +114,12 @@ class CLogicalDynamicIndexGet : public CLogicalDynamicGetBase
 
+	// index scan direction
+	EIndexScanDirection
+	PScanDirection() const
```
Why is it `PScanDirection`? Where does `P` come from?

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/operators/CPhysicalIndexOnlyScan.h:
```diff
@@ -83,6 +86,13 @@ class CPhysicalIndexOnlyScan : public CPhysicalScan
 
+	// index scan direction
+	EIndexScanDirection
+	PIndexScanDirection() const
```
Why `P`?

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformLimit2IndexGet.h:
```diff
@@ -31,12 +32,22 @@ class CXformLimit2IndexGet : public CXformExploration
+													  const IMDIndex *pmdindex);
+
+	// function to validate if indices sort and nulls direction are commutative.
+	static BOOL FAreIndicesCommutative(CBitVector *first_index_props,
```
Remove `Are`

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformLimit2IndexGet.h:
```diff
@@ -31,12 +32,22 @@ class CXformLimit2IndexGet : public CXformExploration
 										   const IMDIndex *pmdindex);
 
+	// function to determine index scan direction given required order spec and index information.
+	static EIndexScanDirection FGetIndexScanDirection(COrderSpec *pos,
```
Why `F`?

gpopt commented on src/backend/gporca/libgpopt/src/operators/CLogical.cpp:
```diff
@@ -186,13 +187,42 @@ CLogical::PosFromIndex(CMemoryPool *mp, const IMDIndex *pmdindex,
+						  IMDType::EcmptL);
+
+			// if nulls direction of key is 0(LAST), choose ENTFirst and vice-versa
+			ent = (pmdindex->KeyNullsDirectionAt(ul) == 0)
```
Please define a macro or const value instead of using literal `0` or `1`.

Sanath97 commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1058,36 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+			{
+				// Sort DESC
+				sort_direction->Replace(i, GPOS_NEW(mp) ULONG(1));
+			}
```
That does make sense, made the change. Thanks for suggestion

gpopt commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1056,6 +1058,45 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+			if (rel_indoption & INDOPTION_DESC)
+			{
+				// Sort DESC: represented by 1
+				sort_direction->Append(GPOS_NEW(mp) ULONG(1));
```
this should also be replaced by macro or constant or enum.

gpopt commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4442,14 +4442,20 @@ CTranslatorDXLToPlStmt::TranslateDXLDynIdxOnlyScan(
-		&index_strategy_list, &index_subtype_list);
+	CDXLNode *index_cond_list_dxlnode = (*dyn_idx_only_scan_dxlnode)
+		[CDXLPhysicalDynamicIndexScan::EdxldisIndexCondition];
+	// Translate Index Conditions if Index isn't used for order by.
```
Can you add a comment about why?

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogical.h:
```diff
@@ -154,9 +161,10 @@ class CLogical : public COperator
+	static COrderSpec *PosFromIndex(
+		CMemoryPool *mp, const IMDIndex *pmdindex, CColRefArray *colref_array,
+		const CTableDescriptor *ptabdesc,
+		EIndexScanDirection scan_direction = EForwardScan);
```
Can we remove default value?

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformUtils.h:
```diff
@@ -186,32 +186,32 @@ class CXformUtils
 		CColRefSet *pcrsScalarExpr, CColRefSet *outer_refs,
 		const IMDIndex *pmdindex, const IMDRelation *pmdrel,
-		BOOL indexForOrderBy = false);
+		EIndexScanDirection indexScanDirection, BOOL indexForOrderBy = false);
```
Can we also remove the default value for indexForOrderBy?

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformUtils.h:
```diff
@@ -446,11 +446,14 @@ class CXformUtils
 						 CColRefSet *pcrsScalarExpr, CColRefSet *outer_refs,
 						 const IMDIndex *pmdindex, const IMDRelation *pmdrel,
-						 BOOL indexForOrderBy = false)
+						 BOOL indexForOrderBy = false,
+						 EIndexScanDirection indexScanDirection =
+							 EForwardScan /*Default direction if unspecified*/)
```
Remove the default value

gpopt commented on src/backend/gporca/libgpopt/src/operators/CLogicalDynamicIndexGet.cpp:
```diff
@@ -249,7 +252,9 @@ CLogicalDynamicIndexGet::OsPrint(IOstream &os) const
-
+	os << ", Index ScanDirection: (";
+	(m_scan_direction == EForwardScan) ? os << "Forward" : os << "Backward";
+	os << ")";
```
output the scan direction only when it is backward. just output
`, Backward Scan`, not `, Index ScanDirection: (Backward)`

gpopt commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -75,6 +75,10 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpopt;
 
+#define SORT_ASC 0
+#define SORT_DESC 1
+#define NULLS_LAST 0
+#define NULLS_FIRST 1
```
There are duplicate defines (in src/backend/gporca/libgpopt/src/operators/CLogical.cpp). 
Can we define them in `gpopt/base/CUtils.h` or other shared place?

gpopt commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -1576,8 +1582,28 @@ CTranslatorExprToDXL::PdxlnDynamicIndexScan(
+		part_mdids = GPOS_NEW(m_mp) IMdIdArray(m_mp);
+		IMdIdArray *current_mdids = popDIS->GetPartitionMdids();
+		// If ScanDirection is backward, reverse the partitions order
+		for (INT i = current_mdids->Size() - 1; i >= 0; i--)
```
What if the partition has desc index, but the query is `order by xx asc limit xx`?

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformJoin2IndexApply.cpp:
```diff
@@ -203,7 +203,7 @@ CXformJoin2IndexApply::CreateAlternativesForBtreeIndex(
 	CExpression *pexprLogicalIndexGet = CXformUtils::PexprLogicalIndexGet(
 		mp, md_accessor, pexprInner, joinOp->UlOpId(), pdrgpexprConjuncts,
-		pcrsScalarExpr, outer_refs, pmdindex, pmdrel);
+		pcrsScalarExpr, outer_refs, pmdindex, pmdrel, false, EForwardScan);
```
add comment param name for `false`, /* the param name*/

Sanath97 replied:
We consider ForwardScan here because, BackwardScan is only supported in the case where we have Order by clause in the query, but this xform handles scenario of a join with some filter conditions and has nothing to do with order by, hence using Forward Scan.

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +223,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+			// If order spec's sort mdid is DESC set req_sort_direction for the key
+			req_sort_direction->ExchangeSet(i);
+		}
+		if (pmdindex->KeySortDirectionAt(i) == 1)
```
use constant

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +223,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+			// If order spec's nulls direction is FIRST set req_nulls_direction for the key
+			req_nulls_direction->ExchangeSet(i);
+		}
+		if (pmdindex->KeyNullsDirectionAt(i) == 1)
```
constant

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +223,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+	return (pos_mdid->Equals(greater_than_mdid) &&
+			pmdindex->KeySortDirectionAt(0) == 1) ||
+				   (!pos_mdid->Equals(greater_than_mdid) &&
+					pmdindex->KeySortDirectionAt(0) == 0)
```
check if you have any other place that is using literal number

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -187,19 +223,115 @@ CXformLimit2IndexGet::FIndexApplicableForOrderBy(
+//        Function to validate if indices sort, nulls direction is commutative.
+//-----------------------------------------------------------------------------
+BOOL
+CXformLimit2IndexGet::FIndicesCommutative(CBitVector *first_index_props,
```
is this really for index commutative?

Sanath97 replied:
Yes, give two indices's sort or nulls directions, this function validates if both indices sort/nulls directions is commutative or not. Rather than naming SortAndNullsCommutative I named `FIndicesCommutative`. Let me know if this answers your question.

gpopt commented:
@soumyadeep2007 We appreciate it if you can have a quick glance from different angles when you have time.

soumyadeep2007 commented:
> @soumyadeep2007 We appreciate it if you can have a quick glance from different angles when you have time.

Sure I will try and do a comparison w/ planner.

soumyadeep2007 commented on src/backend/gporca/libgpopt/src/xforms/CXformSelect2DynamicIndexGet.cpp:
```diff
@@ -125,7 +125,8 @@ CXformSelect2DynamicIndexGet::Transform(CXformContext *pxfctxt,
 			mp, md_accessor, pexprRelational, pexpr->Pop()->UlOpId(), pdrgpexpr,
-			pcrsScalarExpr, nullptr /*outer_refs*/, pmdindex, pmdrel);
+			pcrsScalarExpr, nullptr /*outer_refs*/, pmdindex, pmdrel, false,
+			EForwardScan);
```
What is the reason we are only considering forward scan here?

Sanath97 replied:
We consider ForwardScan here because, BackwardScan is only supported in the case where we have Order by clause in the query, but this xform handles scenario of a filter on top of a partitioned table and has nothing to do with order by, hence using Forward Scan.

soumyadeep2007 commented on src/backend/gporca/libgpopt/src/xforms/CXformSelect2IndexGet.cpp:
```diff
@@ -111,7 +111,8 @@ CXformSelect2IndexGet::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
 			mp, md_accessor, pexprRelational, pexpr->Pop()->UlOpId(), pdrgpexpr,
-			pcrsScalarExpr, nullptr /*outer_refs*/, pmdindex, pmdrel);
+			pcrsScalarExpr, nullptr /*outer_refs*/, pmdindex, pmdrel, false,
+			EForwardScan);
```
What is the reason we are only considering forward scan here?

Sanath97 replied:
We consider ForwardScan here because, BackwardScan is only supported in the case where we have Order by clause in the query, but this xform handles scenario of a filter on top of a regular table and has nothing to do with order by, hence using Forward Scan.

soumyadeep2007 commented on src/backend/gporca/libnaucrates/src/md/CMDIndexGPDB.cpp:
```diff
@@ -321,6 +358,24 @@ CMDIndexGPDB::Serialize(CXMLSerializer *xml_serializer) const
 		available_cols_str);
 	GPOS_DELETE(available_cols_str);
 
+	// Only B-tree indices support sort and nulls direction
+	if (m_index_type == EmdindBtree)
```
Yes currently only B-tree in core will support a sort direction, but there is nothing preventing a new index am from being ordering friendly. Should we check for `amcanorderby` or `amcanorderbyop` instead?

dgkimura replied:
+1

soumyadeep2007 reviewed (commented):
There are a few places where planner is picking backward Index|IndexOnly Scan
and where we aren't in ORCA (grep for "Backward using" in src/test/regress)

aggregates.out - lots of plan diffs here..
Common pattern found in all diffs:
	Planner uses the following gadget for finding max()|min() in queries:
         Initplan
              Limit 1
                  IOScan

create_index_optimizer.out
	- 3 plans where we miss the opportunity in ORCA all deal with an index on
a bool type. For the 1st plan we do a sort->index scan in ORCA and for the other
two we fall back.


IMO we should investigate these plan diffs as part of this PR, or create
follow-ups (as these might be more fundamental). Specially the 1st one: leveraging the index for max/min operations should be a pretty low bar.

24nishant commented on src/backend/gporca/libgpopt/src/operators/CLogical.cpp:
```diff
@@ -186,13 +190,42 @@ CLogical::PosFromIndex(CMemoryPool *mp, const IMDIndex *pmdindex,
+		COrderSpec::ENullTreatment ent = COrderSpec::EntLast;
+		// if scan direction is forward, order spec computed should match the index's
+		// sort and nulls order.
+		if (scan_direction == EForwardScan)
```
Can you please share some reference, on Scan direction vs choice of mdid and ent.

Sanath97 replied:
Sure, in this function `PosFromIndex` we are computing derived order spec, that is the order spec we can satisfy from the indices and their properties in the relation. The value of this derived order spec helps decide if a Sort operator is required in the later stages of optimization. If the derived order spec value is equal to required order spec then, Sort is not enforced, otherwise Sort is enforced to match the required order spec. Value of this derived spec changes based on Scan direction, if scan direction is forward, that means required order and index's order are matching and  order spec computed should match the index's sort and nulls order. If scan direction is backward, that means required order and index's order are commutative and order spec computed should be commutative to the index's sort and nulls order. In both Forward, Backward scan cases we don't need a Sort operator hence we change the value of derived spec to not enforce it. Hope this answers your question.

Sanath97 commented:
> There are a few places where planner is picking backward Index|IndexOnly Scan and where we aren't in ORCA (grep for "Backward using" in src/test/regress)
> 
> aggregates.out - lots of plan diffs here.. Common pattern found in all diffs: Planner uses the following gadget for finding max()|min() in queries: Initplan Limit 1 IOScan
> 
> create_index_optimizer.out - 3 plans where we miss the opportunity in ORCA all deal with an index on a bool type. For the 1st plan we do a sort->index scan in ORCA and for the other two we fall back.
> 
> IMO we should investigate these plan diffs as part of this PR, or create follow-ups (as these might be more fundamental). Specially the 1st one: leveraging the index for max/min operations should be a pretty low bar.

Regarding the first diff from `aggregates.out` where planner leverage Index we already have a follow up task for that and will be picking it up next after this PR. This PR focuses on 
1. Adding the initial support to use Backward index scans wherever required.
2.  Particularly, handle the queries with order by clause and a limit value to leverage index and backward indexscan.
3. Document cases where Backward scans could be used but ORCA currently don't support(Like min/max function). These cases are documented in `qp_indexscan.sql` file, will create follow-up tickets for these cases. The queries in `create_index_optimzier.out` also comes under this category.

Sanath97 commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogical.h:
```diff
@@ -32,6 +32,13 @@ namespace gpopt
 {
 using namespace gpos;
 
+// Scan direction of the index
+enum EIndexScanDirection
+{
+	EBackwardScan = 0,
+	EForwardScan,
+	EisdSentinel
+};
```
> I think we should be matching `EdxlIndexScanDirection` and be considering `NoMovementScanDirection` as well. In planner, `NoMovementScanDirection` is used for indexes that don't have an ordering property.
> 
> This means that for the index: `IndexOptInfo->sortopfamily == NULL` - they are either NOT a btree index or that they don't have a sort operator that maps to the btree opfamily..Reference: `get_relation_info()`)
> 
> Relevant code:
> 
> ```c
> index_is_ordered = (index->sortopfamily != NULL);
> ...
> ipath = create_index_path(root, index,
> 								  index_clauses,
> 								  orderbyclauses,
> 								  orderbyclausecols,
> 								  useful_pathkeys,
> 								  index_is_ordered ?
> 								  ForwardScanDirection :
> 								  NoMovementScanDirection,
> ...
> ```

Regarding `NoMovementScanDirection` we didn't have that as part of the enumerator because, currently only B-tree indices or indices that can support ordering only use IndexScan operator(which uses this enum) all other type of indices that doesn't support ordering use SeqScan with a Sort which doesn't require this enum and hence didn't add NoMovementScanDirection. Hope this answers you question.

Sanath97 commented on src/backend/gporca/libgpopt/src/operators/CLogical.cpp:
```diff
@@ -146,7 +149,8 @@ CLogical::PdrgpdrgpcrCreatePartCols(CMemoryPool *mp, CColRefArray *colref_array,
-					   const CTableDescriptor *ptabdesc)
+					   const CTableDescriptor *ptabdesc,
+					   EIndexScanDirection scan_direction)
 {
 	// compute the order spec after getting the current position of the index key
 	// from the table descriptor. Index keys are relative to the
```
This does make sense to use amcanorder, amcanorderbyop for early return instead of indextype enum, but will note this and make the change as part of a separate PR as this might not come under scope of this current story.

Sanath97 commented:
Pipelines:

https://qpa.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev/jobs/compile_and_test_gpdb/builds/1347
https://qpa.ci.gpdb.pivotal.io/teams/main/pipelines/explain_master

dgkimura reviewed (commented):
I'm still not quite finished reviewing, but leaving another round of comments.

Sanath97 commented on src/backend/gporca/libnaucrates/src/operators/CDXLOperatorFactory.cpp:
```diff
@@ -3721,4 +3721,67 @@ CDXLOperatorFactory::ParseIndexType(const Attributes &attrs)
+
+	ULongPtrArray *ulong_array = GPOS_NEW(mp) ULongPtrArray(mp);
+
+	// Only B-tree indices have sort and nulls directions in dxl
+	// For any other index type, return empty array
+	if (xml_val == nullptr)
+	{
+		for (ULONG ul = 0; ul < num_of_keys; ul++)
+		{
+			ulong_array->Append(GPOS_NEW(mp) ULONG(0));
+		}
+		return ulong_array;
+	}
```
Reason I appended array with 0's is because for any other index type we want to assume ASC(which was case earlier too, before Backwards Index Scan support) as default order as none of the other indices support ordering. The values in this array are accessed later in `PosFromIndex` function using `pmdindex->KeySortDirectionAt(ul)` in `CLogical.cpp`, this function is invoked for all the index types and to avoid `Out of bounds access` issue we use array with 0's rather than returning empty array.

Sanath97 commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2DynamicIndexGet.cpp:
```diff
@@ -0,0 +1,291 @@
+CXformLimit2DynamicIndexGet::FIndexApplicableForOrderBy(
+	CMemoryPool *mp, COrderSpec *pos, const IMDRelation *pmdrel,
+	const IMDIndex *pmdindex, CLogicalDynamicGet *popDynGet)
+{
```
Yes, you are right. Thanks for pointing that case out David. Table with range holes and default partition produces wrong results. Discussed with @gpopt and concluded to not support DynamicIndex(Only) Scans on partition tables as of now. There are a couple reasons for that:

1. Without MergeAppend we could only support DynamicIndex Scans if there exists a btree index on partition column. We cannot support DynamicIndex Scan operator even if btree indices exist on non-partition columns because data for those columns is unordered and spread across multiple partitions. This constraint significantly reduced number of cases this feature could support.
2. Now with the new case of partition table with holes and default partition, number of cases are even more reduced. Moreover, currently ORCA doesn't have information to check if table has a default partition or not. It has to be read from Catalogue tables and fed to ORCA. 
3. Furthermore, ORCA doesn't support composite Partition keys and Multi-Level partitions. Any queries on such tables falls back to Planner.

Considering the significantly reduced cases decided to ignore supporting DynamicIndexScans on partition tables as effort is not worth it.

gpopt replied:
We can open a github issue to track this.

dgkimura commented on src/backend/gporca/libnaucrates/src/md/CMDIndexGPDB.cpp:
```diff
@@ -443,4 +498,58 @@ CMDIndexGPDB::ChildIndexMdids() const
+CWStringDynamic *
+CMDIndexGPDB::SerializeSortAndNullsDirection(CMemoryPool *mp,
+											 ULongPtrArray *dynamic_ptr_array,
+											 bool serialize_Sort) const
```
My comment was more about changing the function signature (not just the name).  If you really want to keep the `serialize_Sort` argument then it would be nice to hear your argument for it.

dgkimura commented on src/backend/gporca/libnaucrates/src/md/CMDIndexGPDB.cpp:
```diff
@@ -250,6 +272,38 @@ CMDIndexGPDB::IncludedColAt(ULONG pos) const
+//
+//	@doc:
+//		Returns n-th column sort direction.
+//		0 corresponds to ASC
+//      1 corresponds to DESC
```
tabs vs spaces

dgkimura commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1106,6 +1109,44 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
 		}
 	}
+
+	// extract sort and nulls direction of the key columns
+	sort_direction = GPOS_NEW(mp) ULongPtrArray(mp);
+	nulls_direction = GPOS_NEW(mp) ULongPtrArray(mp);
```
Not saying it's right or wrong, but what's the rationale for using 2 arrays to represent the index options instead of one array with bitmask similar to what Postgres does?

Sanath97 replied:
I just thought of using 2 arrays because it would be more clear(readable) and easier to manage index options.

dgkimura replied:
Okay, I think that's fair.  The cost is that the space is doubled and any future index options would again increase the size.  Though it hasn't changed yet through PG16.

dgkimura reviewed (commented):
I think I've reviewed everything.  Looks good to me.  Few more minor comments.

Sanath97 commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -4138,4 +4140,53 @@ CXformUtils::FCoverIndex(CMemoryPool *mp, CIndexDescriptor *pindexdesc,
 	return true;
 }
 
+//---------------------------------------------------------------------------
+//	@function:
+//		CXformUtils::GetIndexScanDirection
+//
+//	@doc:
+//		Function to determine index scan direction given required order spec and
+//		index information. This function assumes that caller already validated
+//		if Index Scan is applicable on given index for the required order spec.
+//	    	1. Picks Forward if ORDER BY columns and index keys sort and nulls directions are equal.
+//	    	2. Picks Backward if ORDER BY columns and index keys sort and nulls directions are commutative.
+//---------------------------------------------------------------------------
+EIndexScanDirection
+CXformUtils::GetIndexScanDirection(COrderSpec *pos, const IMDIndex *pmdindex)
+{
+	const CColRef *colref = pos->Pcr(0);
+	IMDId *greater_than_mdid =
+		colref->RetrieveType()->GetMdidForCmpType(IMDType::EcmptG);
+	IMDId *pos_mdid = pos->GetMdIdSortOp(0);
+
+	return (pos_mdid->Equals(greater_than_mdid) &&
+			pmdindex->KeySortDirectionAt(0) == SORT_DESC) ||
+				   (!pos_mdid->Equals(greater_than_mdid) &&
+					pmdindex->KeySortDirectionAt(0) == SORT_ASC)
+			   ? EForwardScan
+			   : EBackwardScan;
+}
```
Yeah, you are right David. Will add comment in the code to make it more apparent.

dgkimura commented on src/backend/gporca/libgpopt/src/xforms/CXformJoin2IndexApply.cpp:
```diff
@@ -201,9 +201,12 @@ CXformJoin2IndexApply::CreateAlternativesForBtreeIndex(
 	CColRefSet *outer_refs, const IMDRelation *pmdrel, const IMDIndex *pmdindex,
 	CXformResult *pxfres)
 {
+	// We consider ForwardScan here because, BackwardScan is only supported in the case where we have Order by clause
```
Wrap line at 80

dgkimura commented on src/backend/gporca/libgpopt/src/xforms/CXformSelect2DynamicIndexGet.cpp:
```diff
@@ -123,9 +123,13 @@ CXformSelect2DynamicIndexGet::Transform(CXformContext *pxfctxt,
 	{
 		IMDId *pmdidIndex = pmdrel->IndexMDidAt(ul);
 		const IMDIndex *pmdindex = md_accessor->RetrieveIndex(pmdidIndex);
+		// We consider ForwardScan here because, because ORCA currently doesn't support BackwardScan on partition tables.
+		// Moreover, BackwardScan is only supported in the case where we have Order by clause in the query, but this xform
+		// handles scenario of a filter on top of a partitioned table.
```
Wrap lines at 80

dgkimura commented on src/backend/gporca/libgpopt/src/xforms/CXformSelect2IndexGet.cpp:
```diff
@@ -109,9 +109,12 @@ CXformSelect2IndexGet::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
 	{
 		IMDId *pmdidIndex = pmdrel->IndexMDidAt(ul);
 		const IMDIndex *pmdindex = md_accessor->RetrieveIndex(pmdidIndex);
+		// We consider ForwardScan here because, BackwardScan is only supported in the case where we have Order by
+		// clause in the query, but this xform handles scenario of a filter on top of a regular table
```
Wrap lines at 80

dgkimura commented on src/backend/gporca/libnaucrates/src/operators/CDXLOperatorFactory.cpp:
```diff
@@ -3721,4 +3721,55 @@ CDXLOperatorFactory::ParseIndexType(const Attributes &attrs)
+ULongPtrArray *
+CDXLOperatorFactory::ExtractConvertBooleanListToULongArray(
+	CDXLMemoryManager *dxl_memory_manager, const XMLCh *xml_val,
+	const XMLCh *true_value, const XMLCh *false_value, ULONG num_of_keys)
```
Add `GPOS_ASSERTS_ONLY` to `true_value` since it is only being used in an assertion.

dgkimura reviewed (approved):
Nice work!  Looks good to me.

dgkimura commented:
We chatted offline, but we didn't call it out explicitly here. Issue raised in comment https://github.com/greenplum-db/gpdb/pull/16087#discussion_r1294975998 should exist in forward dynamic index scan too, right?  If that's the case then I think we should address that ASAP.

gpopt commented:
> We chatted offline, but we didn't call it out explicitly here. Issue raised in comment https://github.com/greenplum-db/gpdb/pull/16087#discussion_r1294975998 should exist in forward dynamic index scan too, right? If that's the case then I think we should address that ASAP.

You are right, we need to address that.

Sanath97 commented:
> We chatted offline, but we didn't call it out explicitly here. Issue raised in comment [#16087 (comment)](https://github.com/greenplum-db/gpdb/pull/16087#discussion_r1294975998) should exist in forward dynamic index scan too, right? If that's the case then I think we should address that ASAP.

Right now ORCA doesn't use(or generate alternatives of) Dynamic Forward Index Scan for ordering on partition tables(which will give wrong results for the partition table with holes and default partition). ORCA only uses Dynamic Seq Scan with Sort operator. But whenever we add support to use Dynamic Forward/Backward Index Scan for partition tables we need to keep in mind to handle this case. We are tracking that with this ticket: https://github.com/greenplum-db/gpdb/issues/16237. Hope this answers your question.

gpopt commented on src/test/regress/expected/qp_indexscan_optimizer.out:
```diff
@@ -579,4 +577,2527 @@ select b from test_index_with_orderby_limit order by b limit 10;
+     e     |   d   |  a   
+-----------+-------+------
+           |       |     
+ col_e9999 | 10000 | 9999
+ col_e9998 |  9999 | 9998
```
The test doesn't help if there are no duplicate values for column e, we won't be able to verify whether d is sorted correctly. And the same applies to other test cases.

gpopt commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,693 @@ set optimizer_enable_indexscan to off;
+
+CREATE TABLE test_index_with_sort_directions_on_orderby_limit (a int, b text, c float, d int, e text, f int);
+-- single col index with default order
+CREATE INDEX dir_index_a on test_index_with_sort_directions_on_orderby_limit using btree(a);
+-- single col index with reverse order
+CREATE INDEX dir_index_b on test_index_with_sort_directions_on_orderby_limit using btree(b desc);
+-- single col index with opp nulls direction
+CREATE INDEX dir_index_c on test_index_with_sort_directions_on_orderby_limit using btree(c nulls first);
+-- multi col index all with all index keys asc
+CREATE INDEX dir_index_bcd on test_index_with_sort_directions_on_orderby_limit using btree(b,c,d);
+-- multi col index all with all index keys desc
+CREATE INDEX dir_index_fde on test_index_with_sort_directions_on_orderby_limit using btree(f desc,d desc,e desc);
+-- multi col index with mixed index keys properties
+CREATE INDEX dir_index_eda on test_index_with_sort_directions_on_orderby_limit using btree(e, d desc nulls last,a);
+-- Covering index with descending and one include column
+CREATE INDEX dir_covering_index_db ON test_index_with_sort_directions_on_orderby_limit(d desc) INCLUDE (b);
```
It is better to move index DDL to just before the test cases, otherwise it is hard to navigate back to find the what is exact index we want to test.

gpopt commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,693 @@ set optimizer_enable_indexscan to off;
+-- Test Case: Test on a regular table with mixed data type columns.
+-- Purpose: Validate if IndexScan with correct scan direction is used on expected index for queries with order by and limit.
+
+CREATE TABLE test_index_with_sort_directions_on_orderby_limit (a int, b text, c float, d int, e text, f int);
```
Please, do change `test_index_with_sort_directions_on_orderby_limit ` to a shorter name, it is so distracting. You can just name it foo or bar, or t, or zoo.

gpopt commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,693 @@ set optimizer_enable_indexscan to off;
+
+CREATE TABLE test_index_with_sort_directions_on_orderby_limit (a int, b text, c float, d int, e text, f int);
+-- single col index with default order
+CREATE INDEX dir_index_a on test_index_with_sort_directions_on_orderby_limit using btree(a);
```
Next time, for simplify, just define index name to `i_a`, or next `i_abc`, because we only care about the columns in the index. too many/long prefix is distracting, especially when we are testing indexes. For this PR, we can just keep the index name as it is.

Sanath97 commented on src/backend/gporca/libgpopt/include/gpopt/base/CUtils.h:
```diff
@@ -35,6 +35,11 @@ namespace gpopt
 
+#define SORT_ASC 0
+#define SORT_DESC 1
+#define NULLS_LAST 0
+#define NULLS_FIRST 1
```
Value of enum COrderSpec::EntFirst and COrderSpec::EntLast is 1 and 2 respectively, whereas the NULLS_LAST and NULLS_FIRST are defined as 0 and 1. We are Serializing/deserializing the sort and nulls direction arrays by writing a new generic function `SerializeBooleanArray`/`ExtractConvertBooleanListToULongArray` which assumes array values to be 0 and 1 since it is boolean, we can always change that values to anything else but then the function might not be useful to any other use case. For this reason, I think it would be better to use these constants than the enum EntFirst, EntLast. Let me know your thoughts.

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -167,12 +178,14 @@ CXformLimit2IndexGet::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
 CXformLimit2IndexGet::FIndexApplicableForOrderBy(
-	COrderSpec *pos, CColRefArray *pdrgpcrIndexColumns,
+	CMemoryPool *mp, COrderSpec *pos, CColRefArray *pdrgpcrIndexColumns,
 	const IMDIndex *pmdindex)
```
This is the suggested changes for this function (not yet tested):
```cpp
EIndexScanDirection
CXformLimit2IndexGet::FIndexApplicableForOrderBy(
	COrderSpec *pos, CColRefArray *pdrgpcrIndexColumns,
	const IMDIndex *pmdindex)
{
	// Ordered IndexScan is only applicable for BTree index
	if (pmdindex->IndexType() != IMDIndex::EmdindBtree)
	{
		return EisdSentinel;
	}
	if (pdrgpcrIndexColumns->Size() < pos->UlSortColumns())
	{
		return EisdSentinel;
	}

	EIndexScanDirection finalDirection = EisdSentinel;

	for (ULONG i = 0; i < pos->UlSortColumns(); i++)
	{
		// ORDER BY columns must match with leading index columns
		const CColRef *colref = pos->Pcr(i);
		if (!CColRef::Equals(colref, (*pdrgpcrIndexColumns)[i]))
		{
			return EisdSentinel;
		}

		ULONG reqOrder = 0;
		ULONG indexOrder = 0;
		IMDId *greater_than_mdid =
			colref->RetrieveType()->GetMdidForCmpType(IMDType::EcmptG);
		if (greater_than_mdid->Equals(pos->GetMdIdSortOp(i)))
		{
			// If order spec's sort mdid is DESC
			reqOrder |= 1<<0;
		}
		if (pos->Ent(i) == COrderSpec::EntFirst)
		{
			// If order spec's nulls direction is FIRST
			reqOrder |= 1<<1;
		}

		if (pmdindex->KeySortDirectionAt(i) == SORT_DESC)
		{
			// If index key's sort direction is DESC
			indexOrder |= 1<<0;
		}
		if (pmdindex->KeyNullsDirectionAt(i) == NULLS_FIRST)
		{
			// If index key's nulls direction is FIRST
			indexOrder |= 1<<1;
		}

		EIndexScanDirection direction;
		if (reqOrder == indexOrder)
		{
			direction = EForwardScan;
		}
		else if ((reqOrder & indexOrder) == 0)
		{
			direction = EBackwardScan;
		}
		else
		{
			return EisdSentinel;
		}

		if (i == 0)
		{
			// first column's scan direction decides the overall direction
			finalDirection = direction;
		}
		else if (finalDirection != direction)
		{
			return EisdSentinel;
		}
	}

	return finalDirection;
}
```

Let me know if I understand your original code correctly or not. :)

Sanath97 replied:
Hmm... I don't think this is gonna work. For instance, consider below cases
1. Lets say index `i_a` is defined as `btree(a)` and user gives the required order as `order by a nulls first`, as per the logic, `indexOrder` value would be 0 and `reqOrder` value would be 2. Since reqOrder & indexOrder (2&0) is 0 we end up choosing BackwardScan, however in this case index scan isn't applicable.
2. Consider index `i_b` defined as `btree(b desc)` and user gives the required order as `order by b nulls first`, as per logic `indexOrder` value would be 1 and `reqOrder` value would be 2. Since reqOrder & indexOrder (2&0) is 0 we end up choosing BackwardScan, however in this case index scan isn't applicable.

There are more cases that might behave undesirably.  I actually tried using an xor(^) instead of and(&) between reqOrder and indexOrder, modified condition to ((reqOrder ^ indexOrder) == 1) but that doesn't seen to work as well.  Let me know your thoughts. Thanks!

gpopt replied:
How about changing
```cpp
(reqOrder & indexOrder) == 0
```
to cpp
```
(reqOrder ^ indexOrder) == 3
```

gpopt replied:
`(reqOrder & indexOrder) == 0` was a clearly wrong loigc.

khannaekta commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,711 @@ set optimizer_enable_indexscan to off;
 explain (costs off) select b from test_index_with_orderby_limit order by b limit 10;
 select b from test_index_with_orderby_limit order by b limit 10;
+-- re-enable indexscan
+set optimizer_enable_indexscan to on;
```
I thought we wanted to do `reset optimizer_enable_indexscan;`

khannaekta commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -1105,6 +1108,44 @@ CTranslatorRelcacheToDXL::RetrieveIndex(CMemoryPool *mp,
+	// Get IndexAmRoutine Struct
+	IndexAmRoutine *am_routine =
+		gpdb::GetIndexAmRoutineFromAmHandler(index_rel->rd_amhandler);
+	index_amcanorder = am_routine->amcanorder;
```
I may have missed this, but why are we using amcanorder only and removed amcanbackward?

Sanath97 replied:
In this if block, we are reading sort & nulls directions values from metadata, sort and nulls directions must be provided if just `amcanorder` is true in the index AM, having sort, nulls direction values does not depend on `amcanbackward` but just depends on `amcanorder`. Moreover, in future there could exist an index AM which has `amcanorder` to true but `amcanbackward` as false, and in that case we might skip reading sort, nulls direction even they exist. This is the reason we are just using `amcanorder`. Hope I am clear. Thanks!

gpopt commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -1592,8 +1598,10 @@ CTranslatorExprToDXL::PdxlnDynamicIndexScan(
-	// TODO: we assume that the index are always forward access.
-
+	// TODO: we assume that the index are always forward access for partition
+	//  tables as ORCA currently doesn't support backward scans on partition
```
change it to one space

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformLimit2IndexGet.h:
```diff
@@ -30,12 +31,14 @@ using namespace gpos;
+	// 1. ORDER BY columns are prefix of the index columns
+	// 2. Sort and Nulls Direction of ORDER BY columns is either equal or
+	//    commutative to the index columns
+	static EIndexScanDirection FIndexApplicableForOrderBy(
```
Function name needs to be updated.

gpopt commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformLimit2IndexGet.h:
```diff
@@ -30,12 +31,14 @@ using namespace gpos;
+	// direction, given OrderSpec and index columns. This function checks if
+	// 1. ORDER BY columns are prefix of the index columns
+	// 2. Sort and Nulls Direction of ORDER BY columns is either equal or
+	//    commutative to the index columns
```
commutative -> reversed

gpopt commented on src/backend/gporca/libgpopt/src/operators/CLogical.cpp:
```diff
@@ -186,13 +187,46 @@ CLogical::PosFromIndex(CMemoryPool *mp, const IMDIndex *pmdindex,
+
+			// if nulls direction of key is 0, choose ENTLast and
+			// vice-versa
+			ent = (pmdindex->KeyNullsDirectionAt(ul) == COrderSpec::EntLast)
+					  ? COrderSpec::EntLast
+					  : COrderSpec::EntFirst;
```
Is the `?:` still needed?

gpopt commented on src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h:
```diff
@@ -118,6 +132,12 @@ class CMDIndexGPDB : public IMDIndex
+	ULONG KeySortDirectionAt(ULONG pos) const override;
+
+	// return the n-th column nulls direction
+	ULONG KeyNullsDirectionAt(ULONG pos) const override;
```
Should we return an enum here?

gpopt commented on src/backend/gporca/libnaucrates/src/md/CMDIndexGPDB.cpp:
```diff
@@ -250,6 +272,38 @@ CMDIndexGPDB::IncludedColAt(ULONG pos) const
+//		CMDIndexGPDB::KeyNullsDirectionAt
+//
+//	@doc:
+//		Returns n-th column Nulls direction
+//		0 corresponds to NULLS LAST
+//		1 corresponds to NULLS FIRST
+//---------------------------------------------------------------------------
+ULONG
```
Do we need to remove the comment? Since we use the enum.

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -163,43 +175,101 @@ CXformLimit2IndexGet::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
+		ULONG indexOrder = 0;
+		IMDId *greater_than_mdid =
+			colref->RetrieveType()->GetMdidForCmpType(IMDType::EcmptG);
+		if (greater_than_mdid->Equals(pos->GetMdIdSortOp(i)))
```
You can just add comment before this line, no need duplicate the comment in every block:
```
// 1st bit represents sort direction, 1 for DESC;
// 2nd bit represents nulls direction, 1 for NULLS FIRST.
```

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -163,43 +175,101 @@ CXformLimit2IndexGet::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
+		if (greater_than_mdid->Equals(pos->GetMdIdSortOp(i)))
+		{
+			// If order spec's sort mdid is DESC
+			// First bit represents Sort direction
```
L222, 228, 235, 241 can be removed.

gpopt commented on src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h:
```diff
@@ -16,6 +16,7 @@
 
 #include "gpos/base.h"
 
+#include "gpopt/base/COrderSpec.h"
```
useless now?

gpopt commented on src/test/regress/sql/qp_indexscan.sql:
```diff
@@ -73,5 +73,711 @@ set optimizer_enable_indexscan to off;
+DROP TABLE part_employee;
+DROP TABLE part_department;
+
+-- Test Case: Union all of two partition tables with order by on their indexed column uses IndexScan's sort property with MergeAppend node
+--            in Planner. However in ORCA we don't generate DynamicIndexScan alternative for union all, also we don't support MergeAppend.
+--            But documenting this case for reference
```
this comment seems not aligned well.

gpopt commented on src/backend/gporca/libgpopt/src/xforms/CXformLimit2IndexGet.cpp:
```diff
@@ -39,7 +39,8 @@ CXformLimit2IndexGet::CXformLimit2IndexGet(CMemoryPool *mp)
 			  GPOS_NEW(mp) CExpression(
-				  mp, GPOS_NEW(mp) CLogicalGet(mp)),  // relational child
+				  mp,
+				  GPOS_NEW(mp) CLogicalGet(mp)),  // relational child
```
I think this diff is not needed.
